### DISCUSSION TITLE
Typed structs using dataclasses

### DISCUSSION THREAD

**Discussion by ratnania** [pyccel/pyccel#1042] [Ideas]
**Describe the bug**
A clear and concise description of what you would like to be implemented.

**Test Code**
Provide code which does not currently work but which should do when this issue is fixed:
```python
Code here
```

**Language**
If appropriate, please specify which language the python code is translated to (Fortran by default)

**Comment by EmilyBourne** [comment]
Is this a duplicate of #773?

**Comment by ratnania** [comment]
This issue is different from #773. 
Let us consider the following Python code
```python
from dataclasses import dataclass
@dataclass
class Point:
    x: int
    y: int
```
We would like to get the following equivalent fortran code
```fortran
type Point
  integer(C_INT64_T) :: x

  integer(C_INT64_T) :: y
end type
```
Actually, I just made this work on the branch devel-843.
For the moment, I had to comment the import
```python
from dataclasses import dataclass
```
**TODO**

1. parse the import and check that the decorator *dataclass* has been imported
2. implement the correponding **c** code
3. make sure that pyccel implements the use a structured type in our pipeline
4. Allow variables to be initialized when defining a structured type; for the moment I'm not using **value** in the **AnnAssign** node of **Python ast**

**Comment by yguclu** [comment]
The `dataclass` decorator simply adds some special methods to a user-defined class: https://docs.python.org/3/library/dataclasses.html.

The problem is that Pyccel does not support user-defined classes yet. We have created an [epic](https://github.com/pyccel/pyccel/issues/453) for them, but it is still empty because we have no concrete plan of action. We should organize a meeting on this topic.

**Comment by ratnania** [comment]
If we consider having structured types in Pyccel, we have different choices:

```python
from dataclasses import dataclass
from typing import NamedTuple

class A:
    x: int
    y: int = 1

@dataclass
class B:
    x: int
    y: int = 1

class C(NamedTuple):
    x: int
    y: int = 1
```
Now, if we create an instance of these classes we get the following results

```python
[in] a = A(x=0)
[out] TypeError: A() takes no arguments

[in] b = B(x=0)
[in] b.y = 3
[in] print(b.x, b.y)
[out] 0 3

[in] c = C(x=0)
[in] c.y = 2
[out] AttributeError: can't set attribute
```

this is why I chose the **dataclass** as it mimics the behavior of a structured type.
